pkg/generator/generate_cert.go

```diff
@@ -0,0 +1,400 @@
+
+	expW, err := time.ParseDuration(expiryW)
+	if err != nil {
+		return fmt.Errorf("failed to expiryWindow %s", err.Error())
+	}
+	c.ExpiryWindow = expW
+
+	return nil
+}
+
+func NewGenerateCert(cl client.Client) GenerateCert {
+	return GenerateCert{
+		client:           cl,
+		CaCertConfig:     &certConfig{},
+		NodeCertConfig:   &certConfig{},
+		ClientCertConfig: &certConfig{},
+	}
+}
+
+// Do func generates the various certificates required and then stores them in respective secrets.
+func (rc *GenerateCert) Do(ctx context.Context, namespace string) error {
+
+	// create the various temporary directories to store the certificates in
+	// the directors will delete when the code is completed.
+	logrus.SetLevel(logrus.InfoLevel)
+
+	certsDir, cleanup := util.CreateTempDir("certsDir")
+	defer cleanup()
+	rc.CertsDir = certsDir
+
+	caDir, cleanupCADir := util.CreateTempDir("caDir")
+	defer cleanupCADir()
+	rc.CAKey = filepath.Join(caDir, "ca.key")
+
+	// generate the base CA cert and key
+	if err := rc.generateCA(ctx, rc.getCASecretName(), namespace); err != nil {
+		msg := " error Generating CA"
+		logrus.Error(err, msg)
+		return errors.Wrap(err, msg)
+	}
+
+	// generate the node certificate for the database to use
+	if err := rc.generateNodeCert(ctx, rc.getNodeSecretName(), namespace); err != nil {
+		msg := " error Generating Node Certificate"
+		logrus.Error(err, msg)
+		return errors.Wrap(err, msg)
+	}
+
+	// generate the client certificates for the database to use
+	if err := rc.generateClientCert(ctx, rc.getClientSecretName(), namespace); err != nil {
+		msg := " error Generating Client Certificate"
+		logrus.Error(err, msg)
+		return errors.Wrap(err, msg)
+	}
+
+	return nil
+}
+
+// generateCA generates the CA key and certificate if not given by the user and stores them in a secret.
+func (rc *GenerateCert) generateCA(ctx context.Context, CASecretName string, namespace string) error {
+
+	// if CA secret is given by user then validate it and use that
+	if rc.CaSecret != "" {
+		logrus.Infof("skipping CA cert generation, using user provided CA secret [%s]", rc.CaSecret)
+
+		secret, err := resource.LoadTLSSecret(CASecretName, resource.NewKubeResource(ctx, rc.client, namespace, kube.DefaultPersister))
+		if err != nil {
+			return errors.Wrap(err, "failed to get CA key secret")
+		}
+
+		// check if the secret contains required info
+		if !secret.ReadyCA() {
+			return errors.Wrap(err, "CA secret doesn't contain the required CA cert/key")
+		}
+
+		if err := ioutil.WriteFile(filepath.Join(rc.CertsDir, resource.CaCert), secret.CA(), security.CertFileMode); err != nil {
+			return errors.Wrap(err, "failed to write CA cert")
+		}
+
+		if err := ioutil.WriteFile(rc.CAKey, secret.CAKey(), security.KeyFileMode); err != nil {
+			return errors.Wrap(err, "failed to write CA key")
+		}
+
+		return nil
+	}
+
+	secret, err := resource.LoadTLSSecret(CASecretName, resource.NewKubeResource(ctx, rc.client, namespace, kube.DefaultPersister))
```

Comment:
Initially we thought of creating these secrets as part of helm-chart and populate them in pre-install job but that is not gonna work because during helm pre hook execution other non-hook resources are not created. So removed empty secret yamls placeholders from the templates and instead our  pre-install job will secret the required secrets. Made required changes in code above approach

This change will leave secrets dangling after helm uninstall is done. This cleanup can he handled in post-delete helm hook